test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs

```diff
@@ -2437,6 +2437,53 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
             }
         }
 
+        [Fact]
+        public async Task ConnectionClosedWhenResponseDoesNotSatisfyMinimumDataRate()
+        {
+            var testContext = new TestServiceContext
+            {
+                SystemClock = new SystemClock(),
+                ServerOptions =
+                {
+                    Limits =
+                    {
+                        MinResponseDataRate = new MinDataRate(bytesPerSecond: double.MaxValue, gracePeriod: TimeSpan.FromSeconds(2))
+                    }
+                }
+            };
+
+            var aborted = new ManualResetEventSlim();
+
+            using (var server = new TestServer(async context =>
+            {
+                context.RequestAborted.Register(() =>
+                {
+                    aborted.Set();
+                });
+
+                var chunk = new string('a', 64 * 1024);
+
+                while (true)
+                {
+                    await context.Response.WriteAsync(chunk, context.RequestAborted);
+                }
+            }, testContext))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    connection.Socket.ReceiveBufferSize = 1;
+
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "Host:",
+                        "",
+                        "");
+
+                    Assert.True(aborted.Wait(TimeSpan.FromSeconds(60)));
```

Comment:
Can we also assert something about the connection state? it should be forcefully closed